systems/controllers/generate_samples.cc

```diff
@@ -0,0 +1,626 @@
+                   drake::systems::Context<AutoDiffXd>* context_ad,
+                   Eigen::VectorXd lcs_state) {
+  // Update autodiff.
+  VectorXd xu_test(n_q + n_v + n_u);
+
+  // u here is set to a vector of 1000s -- TODO why?
+  VectorXd test_u = 1000 * VectorXd::Ones(n_u);
+
+  // Update context with respect to positions and velocities associated with
+  // the candidate state.
+  VectorXd test_q = lcs_state.head(n_q);
+  VectorXd test_v = lcs_state.tail(n_v);
+  xu_test << test_q, test_v, test_u;
+  auto xu_ad_test = drake::math::InitializeAutoDiff(xu_test);
+  plant_ad.SetPositionsAndVelocities(context_ad, xu_ad_test.head(n_q + n_v));
+  multibody::SetInputsIfNew<AutoDiffXd>(plant_ad, xu_ad_test.tail(n_u),
+                                        context_ad);
+
+  plant.SetPositions(context, test_q);
+  plant.SetVelocities(context, test_v);
+  multibody::SetInputsIfNew<double>(plant, test_u, context);
+}
+
+Eigen::VectorXd project_to_surface(
+    Eigen::VectorXd& candidate_state, int min_distance_index,
+    const SamplingC3SamplingParams& sampling_params,
+    drake::multibody::MultibodyPlant<double>& plant,
+    drake::systems::Context<double>* context,
+    const std::vector<
+        std::vector<drake::SortedPair<drake::geometry::GeometryId>>>&
+        contact_geoms) {
+  const auto& query_port = plant.get_geometry_query_input_port();
+  const auto& query_object =
+      query_port.template Eval<drake::geometry::QueryObject<double>>(*context);
+  const auto& inspector = query_object.inspector();
+  SortedPair<GeometryId> verbose_test_pair =
+      contact_geoms.at(0).at(min_distance_index);
+
+  // Get the witness points on each geometry.
+  const SignedDistancePair<double> signed_distance_pair =
+      query_object.ComputeSignedDistancePairClosestPoints(
+          verbose_test_pair.first(), verbose_test_pair.second());
+
+  const Eigen::Vector3d& p_ACa =
+      inspector.GetPoseInFrame(verbose_test_pair.first())
+          .template cast<double>() *
+      signed_distance_pair.p_ACa;
+  const Eigen::Vector3d& p_BCb =
+      inspector.GetPoseInFrame(verbose_test_pair.second())
+          .template cast<double>() *
+      signed_distance_pair.p_BCb;
+
+  // Represent the witness points as points in world frame.
+  RigidTransform T_body1_contact = RigidTransform(p_ACa);
+  const FrameId f1_id = inspector.GetFrameId(verbose_test_pair.first());
+  const Body<double>* body1 = plant.GetBodyFromFrameId(f1_id);
+  RigidTransform T_world_body1 = body1->EvalPoseInWorld(*context);
+  Eigen::Vector3d p_world_contact_a =
+      T_world_body1 * T_body1_contact.translation();
+
+  RigidTransform T_body2_contact = RigidTransform(p_BCb);
+  const FrameId f2_id = inspector.GetFrameId(verbose_test_pair.second());
+  const Body<double>* body2 = plant.GetBodyFromFrameId(f2_id);
+  RigidTransform T_world_body2 = body2->EvalPoseInWorld(*context);
+  Eigen::Vector3d p_world_contact_b =
+      T_world_body2 * T_body2_contact.translation();
```

Comment:
This is a complete reimplementation of `GeomGeomCollider::CalcWitnesPoints`, though implemented less efficiently.  Can you figure out how to use that existing function?

Reply:
Done.

Reply:
Done.